feat: New Personal Access Token doc (#833)
dhis2/dhis2-docs: src/developer/web-api/overview.md

## Authentication { #webapi_authentication } 

> The DHIS2 Web API supports three protocols for authentication, Basic
> Authentication and OAuth 2 and Personal Access Tokens (PAT). You can 
> verify and get information about the currently authenticated user by 
> making a GET request to the following URL:

Reviewer: ```suggestion
The DHIS2 Web API supports three protocols for authentication: 

- [Basic Authentication](#webapi_basic_authentication)
- [Personal Access Tokens (PAT)](#webapi_pat_authentication)
- [OAuth 2](#webapi_oauth2)

You can verify and get information about the currently authenticated 
user by making a GET request to the following URL:
```

a hard requirement in order to provide secure interactions with the Web
API.

## Two-factor authentication { #webapi_2fa } 

DHIS2 supports two-factor authentication. This can be enabled per user.
When enabled, users will be asked to enter a 2FA code when logging in. You
can read more about 2FA [here](https://www.google.com/landing/2step/).

## Personal Access Token { #webapi_pat_authentication }
Personal access tokens (PATs) are an alternative to using passwords for
authentication to DHIS2 when using the API.

PATs can be a more secure alternative to HTTP Basic Authentication,
and should be your preferred choice when creating a new app/script etc. 

> Since HTTP Basic Authentication is considered unsecure because among other things
> that it uses your username and password in clear text, it might be phased out in 
> future versions and made optional, which means it needs to be explicitly enabled 
> in the configuration.

Reviewer: I'd suggest moving this up to the `basic` authentication section and rephrasing it a little:

```suggestion
HTTP Basic Authentication is considered insecure because, among other things, 
it sends your username and password in clear text. It may be deprecated in 
future DHIS2 versions or made opt-in, meaning that basic authentication would 
need to be explicitly enabled in the configuration.
```

Author: Thanks, @amcgee! I agree, this sounds better.

#### Important security concerns!

> Your PAT's will automatically inherit all the permissions/authorizations your
> user has at all time. It is therefore extremely important that you limit the access your
> token has depending on the use-case, see more on how to configure your token.

Reviewer: ```suggestion
Your PATs will automatically inherit all the permissions and authorizations your
user has. It is therefore extremely important that you limit the access granted to
your token depending on how you intend to use it, see **Configuring your token**.
```

**If you only want the token to have access to a narrow and specific part of the
server, it is advised to rather create a new special user that you assign only
the roles/authorities you want it to have access to.**

### Creating a token
To create a new PAT, you have two choices:
* A. Create a token in the UI on your account's profile page.
* B. Create a token via the API

### A. Creating a token on the account's page
Log in with your username and password, go to your profile page
(Click top right corner, and chose "Edit profile" from the dropdown).
On your user profile page, choose "Manage personal access tokens" from the
left side menu.
You should now be on the "Manage personal access tokens" page and see the
text: "You don't have any active personal access tokens".
Click "Generate new token" to make a new token.

> A "Generate new token" popup will be shown and present you with two choices:

Reviewer: We should add a screenshot here, and/or link to separate documentation just for the user profile app

Author: Added some screenshots for the UI flow.

#### 1. Server/script context:
_"This type is used for integrations and scripts that won't be accessed by a browser"._

If you plan to use the token in an application, a script or similar, this
type should be your choice.

#### 2. Browser context:
_"This type us used for applications, like public portals, that will be accessed with a web browser"._

If you need to link to DHIS2 on a webpage, or e.g. embed in an iframe,
this is probably the type of token you want.

### Configuring your token

After choosing what token type you want, you can configure different access constraints on
your token. By constraint, we mean how to limit and narrow down how your token can be used.
This can be of crucial importance if you plan on using the token in a public environment,
e.g. on a public dashboard on another site, embedded in an iframe.
Since tokens always have the same access/authorities that your user currently has, taking special 
care is needed if you intend to use it in any environment you don't have 100% control over.

> **NB**: If anyone else gets the hand on your token, they can do anything your user can, and
> there is no way to see any difference between what is done with the token or your normal user.

Reviewer: ```suggestion
**NB**: If anyone else gets their hands on your token, they can do anything your user can do. 
It is not possible to distinguish between actions performed using the token and other actions
performed by your user.
```

**Important**: It is strongly advised that you create a separate unique user with only the roles/authorities

> you want the token to have if you plan on using PAT tokens in a non-secure and/or public environment,

Reviewer: ```suggestion
you want the token to have if you plan to use PAT tokens in an insecure and/or public environment,
```

e.g. on a PC or server, you don't have 100% control over, or "embedded" in a webpage on another server.

#### The different constraint types are as follows:
* Expiry time

> * Allowed UP addresses

Reviewer: ```suggestion
* Allowed IP addresses
```

* Allowed HTTP methods
* Allowed HTTP referrers

##### Expiry time
Expiry time simply sets for how long you want your token to be usable, the default is 30
days. After the expiry time, the token will simply return a 401 (Unauthorized) message.
You can set any expiry time you want, but it is strongly advised that you set an expiry time 
that is reasonable for your use case.

#### Allowed IP addresses

> This is a comma-separated list of IP addresses you want to limit where the token requests can come from.

Reviewer: ```suggestion
This is a comma-separated list of IP addresses which can send API requests using this Personal Access Token.
```

**Important**: IP address validation relies on the X-Forwarded-For header, which can be spoofed.
For security, make sure a load balancer or reverse proxy overwrites this header.

#### Allowed HTTP methods
A comma-separated list of HTTP methods you want your token to be able to use.
If you only need your token to view data, not modify or delete, selecting only the GET HTTP method 
makes sense.

#### Allowed HTTP referrers

> HTTP referer is a header added to the request, when you click on a link, this says which site/page 

Reviewer: ```suggestion
HTTP referer is a header added to the request by web browsers.  When you click on a link, the browser sets the `referer` header to the site or page 
```

you were on when you clicked the link. 
Read more about the HTTP referer header here: https://en.wikipedia.org/wiki/HTTP_referer

This can be used to limit the use of a "public" token embedded on another page on another site. 

> Making sure that the referer header match the site hostname in should come from, can

Reviewer: ```suggestion
Making sure that the referer header matches the site hostname it should come from can
```

> help you from abuse of the token, e.g. if someone puts it on a forum etc.

Reviewer: ```suggestion
help avoid abuse of the token, e.g. if someone posts it on a public forum.
```

> **Important**: this is not a security feature. The referrer header can easily be spoofed.

Reviewer: ```suggestion
**Important**: this is not a security feature. The `referer` header can easily be spoofed.
```

This setting is intended to discourage unauthorized third-party developers from connecting
to public access instances.

#### Saving your token:
When you are done configuring your token, you can save it by clicking the "Generate new token"
button, on the bottom right of the pop-up.
When doing so the token will be saved and a secret token key will be generated on the server.
The new secret token key will be shown on the bottom of the PAT token list with a green background,
and the text "Newly created token".
The secret token key will look similar to this:
```
d2pat_5xVA12xyUbWNedQxy4ohH77WlxRGVvZZ1151814092
```

> **Important**: This generated secret token key will only be shown once, so it is of importance 

Reviewer: ```suggestion
**Important**: This generated secret token key will only be shown once, so it is important 
```

that you copy the token key now and save it in a secure place for use later. 
The secret token key will be securely hashed on the server, and only the hash of this secret token 
key will be saved to the database. This is done to minimize the security impact if someone gets 

> unauthorized access to the database, this is similar to how passwords are saved in the database.

Reviewer: ```suggestion
unauthorized access to the database, similar to the way passwords are handled.
```

### B. Creating a token via the API

Example of how to create a new Personal Access Token with the API:

```
POST https://play.dhis2.org/dev/api/apiToken
Content-Type: application/json
Authorization: Basic admin district

{}
```
**NB**: Remember the empty JSON body (`{}`) in the payload! 

This will return a response containing a token similar to this:
```json
{
  "httpStatus": "Created",
  "httpStatusCode": 201,
  "status": "OK",
  "response": {
     "responseType": "ApiTokenCreationResponse",
     "key": "d2pat_5xVA12xyUbWNedQxy4ohH77WlxRGVvZZ1151814092",
     "uid": "jJYrtIVP7qU",
     "klass": "org.hisp.dhis.security.apikey.ApiToken",
     "errorReports": []
  }
}
```

> **Important**: The token key will only be showed one here in this response.

Reviewer: ```suggestion
**Important**: The token key will only be shown once here in this response.
```

a hard requirement in order to provide secure interactions with the Web
API.

## Two-factor authentication { #webapi_2fa } 

DHIS2 supports two-factor authentication. This can be enabled per user.
When enabled, users will be asked to enter a 2FA code when logging in. You
can read more about 2FA [here](https://www.google.com/landing/2step/).

## Personal Access Token { #webapi_pat_authentication }
Personal access tokens (PATs) are an alternative to using passwords for
authentication to DHIS2 when using the API.

PATs can be a more secure alternative to HTTP Basic Authentication,
and should be your preferred choice when creating a new app/script etc. 

HTTP Basic Authentication is considered insecure because, among other things, 

> it sends your username and password in clear text. It may be deprecated in 

Reviewer 2: This is not really true though, since the username and password should be encrypted over HTTPS?

Author: Yes, that is true. It does not send it in clear text when using HTTPS, but it will if you are using plain HTTP.
The wording should rather be: “it USES your username and password in cleat text”
I will change it to that.

Reviewer 2: I think we just need to be careful here, since use of a token over HTTP is no more secure than a username and password.  The difference of course being that you can revoke the token, but you can also change your password. 

Use of any authentication method over HTTP is not secure. So, I am not sure that there is a big difference between Basic Authentication and PAT over HTTP...the real problem is the lack of use of HTTPS (which of course should never be the case).

Author: Well, I think that using a PAT token has mainly the benefit that you don't expose your username and password, which can be used to log in normally. The PATs can only be used via the API, and you can limit the access with the constraints. I agree it's not a lot more secure, but I personally think it is the better alternative if we have to choose.
There are more features we want to add to the PATs in the future as, for example, monitoring and a log of the latest uses etc., more constraints in the authorities, as well as a URL pattern constraint.
I think stressing the use of HTTPS and making it maybe “harder” to use the server without, is maybe even more important. Added a tutorial how to set up a server with letsencrypt could be nice anyway. #todo
We also intend to introduce other types of token soon, like JWT for front-end. 
I also think, making another type of more secure authentication alternative “token” for server to server communication should be implemented.

Reviewer 2: Would agree with all of that. We are in the process of revamping the implementation docs, and a more thorough discussion of configuring HTTPS defintely needs to be there.

Completely agree that PATs are the way to go and it sounds like a great feature, but there are still instances in which they are going to have to be used, and just want to be sure we are not giving people a false impression that Basic Authentication is somehow intrinsically insecure, at least over HTTPS when configured correctly.
